### PR TITLE
Fix Issue 16611 - 'Unrecognized type const(void)' error for std.trait…

### PR DESCRIPTION
**Pull request by Darredevil**
…s.fullyQualifiedName

https://issues.dlang.org/show_bug.cgi?id=16611

### CODE REVIEW

**Comment by MetaLang** [comment]
Please add one or more unittests testing this code path.

**Comment by MetaLang** [comment]
@ZombineDev no need for testing for `const void`, `shared void` et al., just use [`std.traits.Unqual`](https://dlang.org/phobos/std_traits.html#Unqual).

**Comment by MetaLang** [comment]
It's a little weird that you can create a type like `const void` or `shared void` in the first place... but it seems that DMD doesn't complain (except when you try to declare a variable of those types, of course).

**Comment by schveiguy** [comment]
@MetaLang need it for declaring appropriate void arrays: `const(void)[]`, `shared(void)[]`, etc.

**Comment by MetaLang** [comment]
Ah, for some reason I thought that was a separate case.

**Comment by Darredevil** [comment]
@ZombineDev @MetaLang Implemented the modifications you suggested. Should be ready for merge once all tests pass.

File: std/traits.d [hunk 1]
```diff
@@ -807,6 +807,9 @@ private template fqnType(T,
         static assert(fqn!(string) == "string");
         static assert(fqn!(wstring) == "wstring");
         static assert(fqn!(dstring) == "dstring");
+        static assert(fqn!(void) == "void");
+        static assert(fqn!(const(void)) == "const(void)");
+        static assert(fqn!(shared(void)) == "shared(void)");
```
Comment (MetaLang) [inline std/traits.d]: Can you also add tests for the following qualifiers:

```d
shared const       void
shared inout       void
shared inout const void
inout              void
inout  const       void
immutable	       void
```

File: std/traits.d [hunk 2]
```diff
@@ -5170,12 +5173,15 @@ enum bool isScalarType(T) = isNumeric!T || isSomeChar!T || isBoolean!T;
...
-enum bool isBasicType(T) = isScalarType!T || is(T == void);
+enum bool isBasicType(T) = isScalarType!T || is(Unqual!T == void);
 
 ///
 @safe unittest
 {
     static assert(isBasicType!void);
+    static assert(isBasicType!(const(void)));
+    static assert(isBasicType!(shared(void)));
+    static assert(isBasicType!(immutable(void)));
```
Comment (MetaLang) [inline std/traits.d]: Ditto.

Comment (Darredevil) [inline std/traits.d]: Done.

**Comment by MetaLang** [comment]
LGTM

**Pull request merged**